File: src/main/java/Day10.kt
class Day10: Day<IntArray> {
    override fun parse(text: String): IntArray {
        return text.split(",").map { it.toInt() }.toIntArray()
    }

    override fun part1(input: IntArray): Any {
        val array = (0..255).toMutableList()
        var currentPosition = 0
        var skip = 0
        for (len in input) {
            reverse(array, currentPosition, len)
            currentPosition += len + skip
            skip++
        }
        return array[0] * array[1]
    }

    private fun reverse(array: MutableList<Int>, currentPosition: Int, len: Int) {
        for (i in 0 until len/2) {
            /*
            * curr i  len   A B
            * 0    0   2    0 1
            * 0    1   2    ---
            * 0    2   23   2  20
            * 2    2   8    4
            * */

            val switchIndexA = (currentPosition + i) % array.size
            val switchIndexB = (currentPosition + len - 1 - i ) % array.size
            val temp = array[switchIndexA]
            array[switchIndexA] = array[switchIndexB]
            array[switchIndexB] = temp
        }
    }

    override fun part2(input: IntArray): Any {
        return 0
    }


}

File: 2019-typescript/src/Day10.ts
import Day from "./Day";
import Map2D from "./Map2D";
import Point from "./Point";

class Day10 implements Day<Map2D<boolean>> {

  parse(text: string): Map2D<boolean> {
    let splitted = text.split('\n')
    let f = function(x: number, y: number): boolean {
      return splitted[y].substring(x, x + 1) === '#'
    }
    let map = new Map2D<boolean>(splitted[0].length, splitted.length, f)
    return map
  }

  findObservations(map: Map2D<boolean>, asteroids: Array<Point>, a: Point): number {
    return asteroids.filter((b: Point) => a !== b).map((b: Point) => {
      let delta = b.minus(a).normalize()
      return map.walk(a, delta, b, (x: boolean) => x) === b ? 1 : 0
    }).reduce((a: number, b: number) => a + b, 0)
  }

  part1(input: Map2D<boolean>): any {
    /*
for each asteroid - a
for each other asteroid - b
dx, dy = distance between a and b
shorten dx, dy
walk in increments of dx, dy from a towards b, if there is any asteroid in the way, return false. otherwise return true.
    */
    let asteroids = input.findPositions((b: boolean) => b)
    let values = asteroids.map((a: Point) => this.findObservations(input, asteroids, a))
    return Math.max(...values)
  }

  part2(input: Map2D<boolean>): any {
    return 0
  }

}

export default Day10;
